internal/exec/copy_glob.go

```diff
@@ -0,0 +1,406 @@
+	for _, pattern := range excluded {
+		matched, err := u.PathMatch(pattern, fullRelPath)
+		if err != nil {
+			log.Debug("Error matching exclusion pattern in prefix function", logKeyPattern, pattern, logKeyPath, fullRelPath, logKeyError, err)
+			continue
+		}
+		if matched {
+			log.Debug("Excluding (prefix) due to exclusion pattern (plain match)", logKeyPath, fullRelPath, logKeyPattern, pattern)
+			return true
+		}
+		if info.IsDir() {
+			matched, err = u.PathMatch(pattern, fullRelPath+"/")
+			if err != nil {
+				log.Debug("Error matching exclusion pattern with trailing slash in prefix function", logKeyPattern, pattern, logKeyPath, fullRelPath+"/", logKeyError, err)
+				continue
+			}
+			if matched {
+				log.Debug("Excluding (prefix) due to exclusion pattern (with trailing slash)", logKeyPath, fullRelPath+"/", logKeyPattern, pattern)
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// processPrefixEntry handles a single entry for copyDirRecursiveWithPrefix.
+func processPrefixEntry(entry os.DirEntry, ctx *PrefixCopyContext) error {
+	fullRelPath := filepath.ToSlash(filepath.Join(ctx.Prefix, entry.Name()))
+	srcPath := filepath.Join(ctx.SrcDir, entry.Name())
+	dstPath := filepath.Join(ctx.DstDir, entry.Name())
+
+	info, err := entry.Info()
+	if err != nil {
+		return fmt.Errorf("getting info for %q: %w", srcPath, err)
+	}
+
+	if entry.Name() == ".git" {
+		log.Debug("Skipping .git directory", logKeyPath, fullRelPath)
+		return nil
+	}
+
+	if shouldSkipPrefixEntry(info, fullRelPath, ctx.Excluded) {
+		return nil
+	}
+
+	if info.IsDir() {
+		if err := os.MkdirAll(dstPath, info.Mode()); err != nil {
+			return fmt.Errorf("creating directory %q: %w", dstPath, err)
+		}
+		newCtx := &PrefixCopyContext{
+			SrcDir:     srcPath,
+			DstDir:     dstPath,
+			GlobalBase: ctx.GlobalBase,
+			Prefix:     fullRelPath,
+			Excluded:   ctx.Excluded,
+		}
+		return copyDirRecursiveWithPrefix(newCtx)
+	}
+	return copyFile(srcPath, dstPath)
+}
+
+// copyDirRecursiveWithPrefix recursively copies srcDir to dstDir while preserving the global relative path.
+func copyDirRecursiveWithPrefix(ctx *PrefixCopyContext) error {
+	entries, err := os.ReadDir(ctx.SrcDir)
+	if err != nil {
+		return fmt.Errorf("reading directory %q: %w", ctx.SrcDir, err)
+	}
+	for _, entry := range entries {
+		if err := processPrefixEntry(entry, ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// getMatchesForPattern returns files/directories matching a pattern relative to sourceDir.
+func getMatchesForPattern(sourceDir, pattern string) ([]string, error) {
+	fullPattern := filepath.Join(sourceDir, pattern)
+	matches, err := u.GetGlobMatches(fullPattern)
+	if err != nil {
+		return nil, fmt.Errorf("error getting glob matches for %q: %w", fullPattern, err)
+	}
+	if len(matches) != 0 {
+		return matches, nil
+	}
+
+	// Handle shallow copy indicator.
+	if strings.HasSuffix(pattern, shallowCopySuffix) {
+		if !strings.HasSuffix(pattern, "/**") {
```

Comment:
We're doing something wrong here if we are doing anything with the wild card symbols. This should be handled by the library.

What about:
- `**{a,b}foo?bar*.txt`